packages/bulkapi/src/database/DbChanges.ts

```diff
@@ -0,0 +1,314 @@
+                values[0].missing
+            )
+        })
+        this.updatesPropertyTable.values().forEach((values: DbPropertyUpdate[]) => {
+            // Just take the first from _values_, as every property except _newValue_ is identical anyway.
+            // And there can only be one newValue as well.
+            const data: PropertyRowData = {
+                node_id: values[0].node_id,
+                property_language: values[0].property.language,
+                property_version: values[0].property.version,
+                property_key: values[0].property.key,
+                value: values[0].newValue
+            }
+            result += this.createQuery(
+                data,
+                "property_language",
+                "property_version",
+                "property_key",
+                PROPERTIES_TABLE,
+                TableHelpers.PROPERTIES_COLUMN_SET,
+                values[0].missing
+            )
+        })
+        // Deletes at the end, so any (useles) upodates on deleted nodes don't give errors.
+        const idsToDelete = this.deletedNodesTable.values().map(v => v[0].id)
+        if (idsToDelete.length > 0) {
+            const sqlIds = sqlArrayFromNodeIdArray(idsToDelete)
+            result += `-- Remove orphans by moving them to the orphan tables
+                DELETE FROM ${NODES_TABLE} n
+                WHERE n.id IN ${sqlIds};
+                
+                DELETE FROM ${PROPERTIES_TABLE} p
+                WHERE p.node_id IN ${sqlIds};
+
+                DELETE FROM ${CONTAINMENTS_TABLE} c
+                WHERE c.node_id IN ${sqlIds};
+
+                DELETE FROM ${REFERENCES_TABLE} r
+                WHERE r.node_id IN ${sqlIds};
+                `
+        }
+        
+        logger.requestLog("DATABASE INSERT " + result)
+        return result
+    }
+
+    /**
+     * Creates a query (update, insert or delete) for a features table.
+     * Generic for properties, references and containments.
+     * @param data
+     * @param languageColum
+     * @param versionColumn
+     * @param keyColum
+     * @param tableName
+     * @param columnSet
+     * @param missing
+     */
+    private createQuery(
+        data: object,
+        languageColum: string,
+        versionColumn: string,
+        keyColum: string,
+        tableName: string,
+        columnSet: ColumnSet,
+        missing: Missing
+    ): string {
+        let result = ""
+        switch (missing) {
+            case Missing.MissingBefore:
+                result += `-- insert new feature for existing node
+                                ${this.context.pgp.helpers.insert(data, columnSet)};`
+                break
+            case Missing.MissingAfter:
+                result += `-- delete feature for existing node
+                                DELETE FROM ${tableName} tabl
+                                WHERE 
+                                    tabl.node_id = '${data["node_id"]}' AND
```

Comment:
How SQL-injectable is this? (Basically same remark as before – maybe can use a `VALUES` clause here as well?)

Reply:
All data is validated.